From 6e1353425b28e00c26bb920c6c25073f35adfec5 Mon Sep 17 00:00:00 2001
From: William Wu <william.wu@rock-chips.com>
Date: Tue, 28 Nov 2017 11:52:51 +0800
Subject: [PATCH] usb: dwc_otg_310: do not stop isoc ep transfer when ep
 dequeue

When dequeue req which already queued on the endpoint,
don't stop the transfer and also don't flush the Tx FIFO
for isoc endpoint, otherwise, it may fail when set NAK
bit and disable ep in dwc_otg_pcd_ep_stop_transfer().

Change-Id: Ib3c1215f9c9507c08cc12c966456209384811b6d
Signed-off-by: William Wu <william.wu@rock-chips.com>
---
 drivers/usb/dwc_otg_310/dwc_otg_pcd.c | 22 ++++++++++++----------
 1 file changed, 12 insertions(+), 10 deletions(-)

diff --git a/drivers/usb/dwc_otg_310/dwc_otg_pcd.c b/drivers/usb/dwc_otg_310/dwc_otg_pcd.c
index ce7263df4556..292f0cae2623 100644
--- a/drivers/usb/dwc_otg_310/dwc_otg_pcd.c
+++ b/drivers/usb/dwc_otg_310/dwc_otg_pcd.c
@@ -2468,16 +2468,18 @@ int dwc_otg_pcd_ep_dequeue(dwc_otg_pcd_t *pcd, void *ep_handle,
 	}
 
 	if (!DWC_CIRCLEQ_EMPTY_ENTRY(req, queue_entry)) {
-		dwc_otg_pcd_ep_stop_transfer(GET_CORE_IF(pcd),
-					     &ep->dwc_ep);
-		/* Flush the Tx FIFO */
-		if (ep->dwc_ep.is_in) {
-			dwc_otg_flush_tx_fifo(GET_CORE_IF(pcd),
-					      ep->dwc_ep.tx_fifo_num);
-			release_perio_tx_fifo(GET_CORE_IF(pcd),
-					      ep->dwc_ep.tx_fifo_num);
-			release_tx_fifo(GET_CORE_IF(pcd),
-					ep->dwc_ep.tx_fifo_num);
+		if (ep->dwc_ep.type != DWC_OTG_EP_TYPE_ISOC) {
+			dwc_otg_pcd_ep_stop_transfer(GET_CORE_IF(pcd),
+						     &ep->dwc_ep);
+			/* Flush the Tx FIFO */
+			if (ep->dwc_ep.is_in) {
+				dwc_otg_flush_tx_fifo(GET_CORE_IF(pcd),
+						      ep->dwc_ep.tx_fifo_num);
+				release_perio_tx_fifo(GET_CORE_IF(pcd),
+						      ep->dwc_ep.tx_fifo_num);
+				release_tx_fifo(GET_CORE_IF(pcd),
+						ep->dwc_ep.tx_fifo_num);
+			}
 		}
 
 		dwc_otg_request_done(ep, req, -DWC_E_RESTART);
-- 
2.35.3

